### PR TITLE
some trivial nspawn related fixes

### LINKED ISSUE #8194: systemd-nspawn -u foo fails if foo is not a valid user on the host system
`systemd-nspawn -u` is documented as "After transitioning into the container, change to the specified user-defined in the container's user database.". In systemd 234, this works fine even if the user specified doesn't exist on the host system, as long as it's available within the container. This changed in 237, where nspawn started failing explicitly in this case with a `Failed to resolve user` error. We've noticed this when trying to use `mock`, as internally that calls `systemd-nspawn -u mockbuild`, where `mockbuild` doesn't exist on the host (by design), but it defined as `mockbuild:x:0:156::/builddir:/bin/bash` in `/etc/passwd` in the mock chroot. @zfjagann and I did some digging and think this might be related to 7d4904fe7a (which was part of #7767), but we haven't been able to validate that as the commit doesn't revert cleanly in the released 237 tree.

### Submission type

 - Bug report

### systemd version the issue has been seen with

> 237

### Used distribution

> CentOS 7

### In case of bug report: Expected behaviour you didn't see

> nspawn should start the container and run the command as the specified user

### In case of bug report: Unexpected behaviour you saw

> `Failed to resolve user mockbuild.`

### In case of bug report: Steps to reproduce the problem

> Minimal repro: given a mock chroot on `/data/mock/centos-7-x86_64`, run
`systemd-nspawn -q -D /data/mock/centos-7-x86_64 -u mockbuild bash --login -c '/usr/bin/cat /etc/passwd'`
which is a simplified version of what `mock` runs internally. On 234, this produces the contents of `/etc/passwd` in the container as expected. On 237, it results in `Failed to resolve user mockbuild.`.

### PR DESCRIPTION
**Pull request by poettering**
I tried to fix #8194, but couldn't reproduce it, hence I reworked some stuff I noticed on the way could use some reworking...

### CODE REVIEW

File: src/nspawn/nspawn-setuid.c [hunk 1]
```diff
@@ -118,21 +120,18 @@ int change_uid_gid(const char *user, char **_home) {
...
         fd = -1;
 
-        if (!fgets(line, sizeof(line), f)) {
-                if (!ferror(f)) {
-                        log_error("Failed to resolve user %s.", user);
-                        return -ESRCH;
-                }
-
-                return log_error_errno(errno, "Failed to read from getent: %m");
+        r = read_line(f, LONG_LINE_MAX, &line);
```
Comment (keszybz) [inline src/nspawn/nspawn-setuid.c:128]: We could use a lower limit here, not that it matters in practice.

File: src/basic/fd-util.h [hunk 2]
```diff
@@ -35,6 +35,13 @@ int close_nointr(int fd);
 int safe_close(int fd);
 void safe_close_pair(int p[]);
 
+static inline int safe_close_above_stdio(int fd) {
+        if (fd < 3) /* keep stdin/stdout/stderr open + propagate closed as closed. */
```
Comment (keszybz) [inline src/basic/fd-util.h]: Some callers would benefit from -1 being returned unconditionally, and others wouldn't care. Maybe this should be changed?

File: src/basic/process-util.c [hunk 3]
```diff
@@ -1186,10 +1187,14 @@ pid_t getpid_cached(void) {
 
                 new_pid = raw_getpid();
 
-                if (__register_atfork(NULL, NULL, reset_cached_pid, __dso_handle) != 0) {
-                        /* OOM? Let's try again later */
-                        cached_pid = CACHED_PID_UNSET;
-                        return new_pid;
+                if (!installed) {
+                        if (__register_atfork(NULL, NULL, reset_cached_pid, __dso_handle) != 0) {
```
Comment (keszybz) [inline src/basic/process-util.c]: FTR: it returns either 0 or -ENOMEM, so this conditional is correct.

**Comment by keszybz** [comment]
I'll set the needs-rework flag, even though it doesn't strictly apply, so that it's visible in the listing that it's waiting for response.

File: src/basic/process-util.c (hunk 3 above)
Comment (poettering) [reply src/basic/process-util.c]: given it's semi-documented state (glibc doesn't document it, and LSB only very superficially) I didn't want to hardcode errors, hence I tried to make the error check as generic as possible, which is why I didn't want to check the actual numbers

Comment (poettering) [reply src/basic/process-util.c]: (but i figure i could add this as a comment)

**Comment by poettering** [comment]
Force pushed a new version with all points addressed. No other changes. PTAL!

**Comment by poettering** [comment]
Force pushed a new rebased version now. No other changes than the rebase fix.

**Comment by poettering** [comment]
Took the liberty to put the green label on, given that you liked it already and I only rebased the PR without any other changes.

**Pull request merged**